### ISSUE TITLE
completely wrong error message

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#936]
This image is of a screen-capture, sorry. But I think it make the point clearly. The real bug is that I used `.` instead of `!` to access a `ref` field. But look at the testing output error! This at least got a really good laugh from the class (and maybe also slightly sent the message, “Don't actually read the output of Pyret”).

![image](https://cloud.githubusercontent.com/assets/75030/20351743/52d90fe4-abe2-11e6-9135-5d3650260e0f.png)

**Comment by blerner** [comment]
This reaction on your part is rather smug and incorrect, actually. If you mouse over the 1 and 1 in the error message, one of them pops up a hazard triangle and a tooltip saying, "May be stale, click to refresh" and the other one doesn't. Because, duh, they're different data. (And mousing over data is something we've already trained students to do, so they can go clicking on shrunken images, fractions, data values, objects, etc.)

But *you* were the one who objected to rendering reference fields with a wrapper, as opposed to implicitly referring their contents, because you'd wanted to teach cyclic data without having to teach references first. So the current ui is a compromise, one that I've asked for feedback and improvements on repeatedly. The fact that you don't like it now puts the burden on you to either resolve your pedagogy or suggest a better ui that you could explain more easily.

**Comment by jpolitz** [comment]
IMO, the screenshot is a bug.  Printing pixel-identical values, reporting them as different, and saying "you have to mouse over it to see why" isn't OK for _the main case where a mistake could be made_ .  (There's probably plenty of constructible corner cases where you can get get identical renderings for non-equal values, but they are curiosities, not the thing many folks might try first.). The fact that we've asked for feedback doesn't stop it from being a bug.

I agree that this pedagogy and the semantics of references has been far from settled for a long time, though.

Looking at this now, with the current lookup semantics, I think this should probably print as `ref(1)`, and when a data value with mutable fields prints, it should but the `ref` keyword in front of each mutable field value, in order to clearly distinguish in text as well as in interaction.

Either that or looking up a mutable field should be an error.  That's a language design discussion that should happen at some point.

**Comment by blerner** [comment]
No, I disagree - the screenshot is not the bug.  The bug is somewhere in the mix of our semantics for refs, our semantics for dot access, and our design choices for printing values.  The _error message_ is not "completely wrong" at all.  This will require either a breaking change in semantics or a breaking design change in how we render values to fix; it does _not_ require fixing the error message.

**Comment by jpolitz** [comment]
OK, we can split hairs.  The screenshot is a _symptom_ of a bug somewhere.

**Comment by shriram** [comment]
Identical to https://github.com/brownplt/pyret-lang/issues/1241 BUT please make sure that when this is changed, all the examples in that post are also fixed (the internal errors, the identical renderings, etc.).

**Comment by sorawee** [comment]
Can we use background colors to signify `ref`?

**Comment by blerner** [comment]
Not in general, since you could have a `box(square(40, "solid", "red"))` that hides the background.  We could *maybe* use a border here, but that might look like it's part of the image.  You could also have a `box(some massive table)`, which has its own row/column highlighting, etc.

One possibility I hadn't considered before is that we highlight the entire REPL interaction, rather than the individual values, at least at first.  So for instance, if we render *any* ref cells in an interaction, then we immediately and permanently display the warning icon beside the entire interaction.  The icon would have a tooltip, "WARNING: Some values in this rendering are mutable and may be stale.  Click to toggle highlighting them", and clicking on that icon would add a border to the rendered refs.  Note that if we do this, we've taken away the click-to-rerender interaction, but I don't know if that's doing anyone any good anyway.
